### PR TITLE
Better symmetry reasoning in CAAT

### PR DESCRIPTION
**Pull request by ThomasHaas**
This PR improves symmetry reasoning in CAAT for benchmarks with large symmetries.

EDIT: To streamline some code and avoid special corner cases, I also added a `NormalizeFenceRelations` pass that rewrites `fencerel(F)` to `po;[F];po`. CAAT no longer supports the standard `fencerel`.

### CODE REVIEW

File: dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java [hunk 1]
```diff
@@ -496,6 +495,10 @@ private static RefinementModel generateRefinementModel(Wmm original) {
                 if (name.equals(DATA) || name.equals(CTRL) || name.equals(ADDR) || name.equals(CRIT)) {
                     constraintsToCut.add(c);
                 }
+            } else if (c instanceof Definition def && def instanceof Fences) {
+                // (iii) continued: fencerel(F) is unsupported in CAAT
+                //  It should get rewritten to "po;[F];po" but if it was not, we cut it instead.
```
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java:500]: "It should be rewritten" -> "It should be rewritten by the NormalizeFenceRelations pass"

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/processing/NormalizeFenceRelations.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    public static NormalizeFenceRelations newInstance() {
+        return new NormalizeFenceRelations();
+    }
+
+    @Override
+    public void run(Wmm wmm) {
+
+        final Relation po = wmm.getRelations()
+                .stream().filter(r -> r.getDefinition() instanceof ProgramOrder)
+                .findFirst().orElseThrow();
```
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/wmm/processing/NormalizeFenceRelations.java:29]: why not simply `wmm.getRelation(PO)`?

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/wmm/processing/NormalizeFenceRelations.java:29]: Cause I was stupid...

File: dartagnan/src/main/java/com/dat3m/dartagnan/solver/caat4wmm/coreReasoning/CoreReasoner.java [hunk 3]
```diff
@@ -4,106 +4,170 @@
...
+
+        // (2) Remove dominated core reasons and sort by reason length
+        List<Conjunction<CoreLiteral>> reducedCoreReasons = new ArrayList<>(new DNF<>(base2core.values()).getCubes());
+        reducedCoreReasons.sort(Comparator.comparingInt(Conjunction::getSize));
+
+        // (3) Translate back to base reasons (we need the original base reasons for symmetry reasoning)
+        List<Conjunction<CAATLiteral>> reducedBaseReasons = reducedCoreReasons.stream().map(core2base::get).toList();
+
+        // (4) Recompute core reasons with symmetry reasoning.
+        //  Stop early, if the number of reasons gets too large.
```
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/solver/caat4wmm/coreReasoning/CoreReasoner.java:69]: Please add a comment that this does not affect soundness

File: dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java (hunk 1 above)
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java:500]: I will add that it should get rewritten by "some pass". If I add the concrete pass name, we will 100% forget to update it if the pass name ever changes :P.

**Review by hernanponcedeleon — APPROVED** [review]

**Pull request merged**